File: Cores/OSOL.Extremum.Core.DotNet/CodeFeatures/FunctionParser/Parser.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using Newtonsoft.Json.Linq;
using OSOL.Extremum.Core.DotNet.Arithmetics;
using OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.Trees;

namespace OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser
{
    public static class Parser
    {
        public class UnsupportedOperation: Exception {}
        
        private static FileStream ExtractResource(string filename, string suffix, string location = "", string where = ".")
        {
            var assembly = Assembly.GetExecutingAssembly();
            string resourceName = $"OSOL.Extremum.Core.DotNet.Resources.{filename}.{suffix}";
            if(location.Length != 0)
            {
                resourceName = $"OSOL.Extremum.Core.DotNet.Resources.{location}.{filename}.{suffix}";
            }
            var resource = assembly.GetManifestResourceStream(resourceName);

            if (!Directory.Exists(where))
            {
                Directory.CreateDirectory(where);
            }

            var file = File.Create($"{where}/{filename}.{suffix}");
            resource.CopyTo(file);
            file.Close();

            return file;
        }
        
        public static JObject ParseString(string str)
        {
            var rootFolder = "temp";
            var parserLibFile = ExtractResource("parser", "py", "parser", $"{rootFolder}/parser");
            var parserLibInitFile = ExtractResource("__init__", "py", "parser", $"{rootFolder}/parser");
            var parserAppFile = ExtractResource("parser_app", "py", "apps", rootFolder);
            
            
            
            ProcessStartInfo start = new ProcessStartInfo();
            start.FileName = "python";
            start.Arguments = $"{parserAppFile.Name} --function \"{str}\"";
            start.UseShellExecute = false;
            start.RedirectStandardOutput = true;
            
            using (Process process = Process.Start(start))
            {
                using (StreamReader reader = process.StandardOutput)
                {
                    string result = reader.ReadToEnd().Replace('\'', '\"');
                    
                    Directory.Delete(rootFolder, true);
                    return JObject.Parse(result);
                }
            }
        }

        public static Tree<double> BuildTreeD(JObject json)
        {
            switch (json["type"].Value<string>())
            {
                case "binary":
                {
                    var op = json["op"].Value<string>();
                    var leftTree = BuildTreeD((JObject)json["left"]);
                    var rightTree = BuildTreeD((JObject)json["right"]);
                    switch (op)
                    {
                        case "add": return new TreeD.AdditionTree(leftTree, rightTree);
                        case "sub": return new TreeD.SubtractionTree(leftTree, rightTree);
                        case "mult": return new TreeD.MultiplicationTree(leftTree, rightTree);
                        case "div": return new TreeD.DivisionTree(leftTree, rightTree);
                        case "pow": return new TreeD.PowerTree(leftTree, rightTree);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "unary":
                {
                    var op = json["op"].Value<string>();
                    var operandTree = BuildTreeD((JObject)json["operand"]);
                    switch (op)
                    {
                        case "usub": return new TreeD.NegTree(operandTree);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "func":
                {
                    var op = json["func"].Value<string>();
                    var argsTrees = BuildTreeD((JObject)((JArray)json["args"])[0]);
                    switch (op)
                    {
                        case "sin": return new TreeD.SinTree(argsTrees);
                        case "cos": return new TreeD.CosTree(argsTrees);
                        case "exp": return new TreeD.ExpTree(argsTrees);
                        case "abs": return new TreeD.AbsTree(argsTrees);
                        case "ln": return new TreeD.LnTree(argsTrees);
                        case "sqrt": return new TreeD.SqrtTree(argsTrees);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "const": return new TreeD.ConstantTree(json["value"].Value<double>());
                case "var": return new TreeD.VariableTree(json["name"].Value<string>());
                default: throw new SerializationException();
            }
        }
        
        public static Tree<Interval> BuildTreeI(JObject json)
        {
            switch (json["type"].Value<string>())
            {
                case "binary":
                {
                    var op = json["op"].Value<string>();
                    var leftTree = BuildTreeI((JObject)json["left"]);
                    var rightTree = BuildTreeI((JObject)json["right"]);
                    switch (op)
                    {
                        case "add": return new TreeI.AdditionTree(leftTree, rightTree);
                        case "sub": return new TreeI.SubtractionTree(leftTree, rightTree);
                        case "mult": return new TreeI.MultiplicationTree(leftTree, rightTree);
                        case "div": return new TreeI.DivisionTree(leftTree, rightTree);
                        case "pow": return new TreeI.PowerTree(leftTree, rightTree);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "unary":
                {
                    var op = json["op"].Value<string>();
                    var operandTree = BuildTreeI((JObject)json["operand"]);
                    switch (op)
                    {
                        case "usub": return new TreeI.NegTree(operandTree);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "func":
                {
                    var op = json["func"].Value<string>();
                    var argsTrees = BuildTreeI((JObject)((JArray)json["args"])[0]);
                    switch (op)
                    {
                        case "sin": return new TreeI.SinTree(argsTrees);
                        case "cos": return new TreeI.CosTree(argsTrees);
                        case "exp": return new TreeI.ExpTree(argsTrees);
                        case "abs": return new TreeI.AbsTree(argsTrees);
                        case "ln": return new TreeI.LnTree(argsTrees);
                        case "sqrt": return new TreeI.SqrtTree(argsTrees);
                        default: throw new UnsupportedOperation();
                    }
                    break;
                }
                case "const": return new TreeI.ConstantTree(json["value"].Value<double>());
                case "var": return new TreeI.VariableTree(json["name"].Value<string>());
                default: throw new SerializationException();
            }
        }
    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/CodeFeatures/FunctionParser/Parser.scala
package OSOL.Extremum.Cores.JVM.CodeFeatures.FunctionParser

import java.io.{File, PrintWriter}

import OSOL.Extremum.Cores.JVM.Arithmetics.Interval
import OSOL.Extremum.Cores.JVM.CodeFeatures.FunctionParser.Trees.{Tree, TreeD, TreeI}
import OSOL.Extremum.Cores.JVM.CodeFeatures._
import OSOL.Extremum.Cores.JVM.CodeFeatures.FunctionParser.Trees.{Tree, TreeI}
import spray.json._

import scala.sys.process.Process

object Parser {
  
  class UnsupportedOperation(message: String) extends Exception

  def buildTreeD(json: JsValue): Tree[Double] = {
    val jsonObject = json.asJsObject
    jsonObject.getFields("type") match {
      case Seq(JsString("binary")) => {
        val Seq(op, left, right) = jsonObject.getFields("op", "left", "right")
        val leftTree = buildTreeD(left)
        val rightTree = buildTreeD(right)
        op match {
          case JsString("add") => new TreeD.AdditionTree(leftTree, rightTree)
          case JsString("sub") => new TreeD.SubtractionTree(leftTree, rightTree)
          case JsString("mult") => new TreeD.MultiplicationTree(leftTree, rightTree)
          case JsString("div") => new TreeD.DivisionTree(leftTree, rightTree)
          case JsString("pow") => new TreeD.PowerTree(leftTree, rightTree)
          case _ => throw new UnsupportedOperation(s"Unsupported op: $op")
        }
      }
      case Seq(JsString("unary")) => {
        val Seq(op, operand) = jsonObject.getFields("op", "operand")
        val operandTree = buildTreeD(operand)
        op match {
          case JsString("usub") => new TreeD.NegTree(operandTree)
          case _ => throw new UnsupportedOperation(s"Unsupported op: $op")
        }
      }
      case Seq(JsString("func")) => {
        val Seq(func, args) = jsonObject.getFields("func", "args")
        val argsTrees = args.asInstanceOf[JsArray].elements.map(buildTreeD)
        func match {
          case JsString("sin") => new TreeD.SinTree(argsTrees(0))
          case JsString("cos") => new TreeD.CosTree(argsTrees(0))
          case JsString("exp") => new TreeD.ExpTree(argsTrees(0))
          case JsString("abs") => new TreeD.AbsTree(argsTrees(0))
          case JsString("ln") => new TreeD.LnTree(argsTrees(0))
          case JsString("sqrt") => new TreeD.SqrtTree(argsTrees(0))
          case _ => throw new UnsupportedOperation(s"Unsupported func: $func")
        }
      }
      case Seq(JsString("const")) => {
        val Seq(value) = jsonObject.getFields("value")
        new TreeD.ConstantTree(value.toString().toDouble)
      }
      case Seq(JsString("var")) => {
        val Seq(name) = jsonObject.getFields("name")
        new TreeD.VariableTree(name.toString().drop(1).dropRight(1))
      }
      case _ => throw new DeserializationException(s"Can't parse $json")
    }
  }

  def buildTreeI(json: JsValue): Tree[Interval] = {
    val jsonObject = json.asJsObject
    jsonObject.getFields("type") match {
      case Seq(JsString("binary")) => {
        val Seq(op, left, right) = jsonObject.getFields("op", "left", "right")
        val leftTree = buildTreeI(left)
        val rightTree = buildTreeI(right)
        op match {
          case JsString("add") => new TreeI.AdditionTree(leftTree, rightTree)
          case JsString("sub") => new TreeI.SubtractionTree(leftTree, rightTree)
          case JsString("mult") => new TreeI.MultiplicationTree(leftTree, rightTree)
          case JsString("div") => new TreeI.DivisionTree(leftTree, rightTree)
          case JsString("pow") => new TreeI.PowerTree(leftTree, rightTree)
          case _ => throw new UnsupportedOperation(s"Unsupported op: $op")
        }
      }
      case Seq(JsString("unary")) => {
        val Seq(op, operand) = jsonObject.getFields("op", "operand")
        val operandTree = buildTreeI(operand)
        op match {
          case JsString("usub") => new TreeI.NegTree(operandTree)
          case _ => throw new UnsupportedOperation(s"Unsupported op: $op")
        }
      }
      case Seq(JsString("func")) => {
        val Seq(func, args) = jsonObject.getFields("func", "args")
        val argsTrees = args.asInstanceOf[JsArray].elements.map(buildTreeI)
        func match {
          case JsString("sin") => new TreeI.SinTree(argsTrees(0))
          case JsString("cos") => new TreeI.CosTree(argsTrees(0))
          case JsString("exp") => new TreeI.ExpTree(argsTrees(0))
          case JsString("abs") => new TreeI.AbsTree(argsTrees(0))
          case JsString("ln") => new TreeI.LnTree(argsTrees(0))
          case JsString("sqrt") => new TreeI.SqrtTree(argsTrees(0))
          case _ => throw new UnsupportedOperation(s"Unsupported func: $func")
        }
      }
      case Seq(JsString("const")) => {
        val Seq(value) = jsonObject.getFields("value")
        new TreeI.ConstantTree(Interval(value.toString().toDouble))
      }
      case Seq(JsString("var")) => {
        val Seq(name) = jsonObject.getFields("name")
        new TreeI.VariableTree(name.toString().drop(1).dropRight(1))
      }
      case _ => throw new DeserializationException(s"Can't parse $json")
    }
  }

  def parseString(str: String): JsValue = {
    val rootFolder = "temp"
    val parserLibFile = extractResource("parser", "py", "parser", s"$rootFolder/parser")
    val parserLibInitFile = extractResource("__init__", "py", "parser", s"$rootFolder/parser")
    val parserAppFile = extractResource("parser_app", "py", "apps", s"$rootFolder")
    val treeJson = (Process("python",  Seq(parserAppFile.getAbsolutePath, "--function", str)) !!)
      .replace('\'', '\"')
      .replace("u\"", "\"")
      .parseJson
    deleteDirectory(new File(rootFolder))
    treeJson
  }

  def parseToDoubleTree(str: String): Tree[Double] = str |> parseString |> buildTreeD

  def parseToIntervalTree(str: String): Tree[Interval] = str |> parseString |> buildTreeI
}